Ikiwiki wikilinks

csrhodes opened the pull request:
Superseding closed pull request #2333: I have hooked it into the syntax extensions (off by default) and written minimal documentation.  (Sorry if this is duplication of information; I don't know if you get notified about comments on a closed pull request).

csrhodes committed:
hacky treatment of ikiwiki wikilinks

rewrites them to be standard links to an id with name generated from
the textual form of the wikilink.

This is problematic in that the exact form of the
wikilink (e.g. relative vs absolute) will give a different resulting
name; some consistency will be required in authoring to make
everything work.  However, it works well enough to turn multiple
ikiwiki pages into a single document with internal links.

csrhodes committed:
fix for unused raw variable

csrhodes committed:
make ikiwiki_wikilinks a non-default extension

csrhodes committed:
give wikilinks a title of "wikilink"

csrhodes committed:
round-tripping ikiwiki wikilinks

Since we now know (FSVO "know") what links resulted from wikilinks, we
can write them back out using ikiwiki wikilink syntax (assuming the
extension is turned on).

In order to get round-tripping to work, we no longer do any manipulation
of the link destination.  This is perhaps an unfortunate default, but
there's no other way to assure being able to round-trip.  In practice,
this means that filters will need to be used when exporting to other
formats in order to massage link destinations appropriately.

csrhodes commented:
With the new commits, I am able to get the effect I want for my current project with a filter like https://gist.github.com/csrhodes/cb77a15aa968e6cd0886, and markdown with ikiwiki links can now round-trip.

csrhodes committed:
test cases

csrhodes commented:
I've written some test cases for the ikiwiki_wikilinks functionality.

Please let me know if there's anything else I can do to help get this integrated.

csrhodes commented:
I've rebased to the new code (with `Link` objects having attributes in the constructors).  The tests pass and (after fighting a bit with stack) my filter-based munging seems to do the right thing.

csrhodes commented:
I've fixed the accumulated merge conflicts, and checked that the test cases continue to pass (and that my actual application works, too).

Is there any chance of merging this support for ikiwiki-style wikilinks?

Thanks,

Christophe

jgm commented:
``` haskell
+          srcstr = (drop (idx + 1) rawstr) ++ "\n"
...
+      src <- parseFromString (mconcat <$> many (count 1 anyChar)) (init srcstr)
```
This makes me worry about the rest.  First, doesn't `mconcat <$> many (count 1 anyChar)` just amount to a highly inefficient version of `many anyChar` or for that matter `getContents`?  Second, what is the point of this `parseFromString`  Isn't `src` just going to be `init srcstr` in the end?  Third, why do you add the newline to the end of `srcstr` if you're just going to ignore it later using `init`?

csrhodes commented:
@jgm Thanks for the feedback.  This is just a note to say that you are certainly right to worry; I might have said in my initial pull request that this is my first attempt at Haskell of any kind, but if not, it's probably obvious from the code.  Honestly, on the specific issue, what I've probably done is used the nearest hammer to pacify the nail of the type system.

I can attempt to simplify and rewrite to the extent that I understand what I'm doing.  My own worry is that without direction I might make things worse, or at least not better enough, and I don't want to waste your time with updated pull requests that aren't noticeably better from your perspective.

Thanks,

Christophe

csrhodes committed:
slightly less bad code

Following initial feedback from jgm at
<https://github.com/jgm/pandoc/pull/2342#issuecomment-230058594>:

- init is not "initialize for parsing", it's "butlast"
- mconcat <$> many (count 1 anyChar) is (many anyChar)

so fix those.

jgm commented:
I'll be happy to give more feedback -- but I'm pretty short
of time at the moment, so it might have to wait a bit.

+++ Christophe Rhodes [Jul 04 16 12:58 ]:
>   [1]@jgm Thanks for the feedback. This is just a note to say that you
>   are certainly right to worry; I might have said in my initial pull
>   request that this is my first attempt at Haskell of any kind, but if
>   not, it's probably obvious from the code. Honestly, on the specific
>   issue, what I've probably done is used the nearest hammer to pacify the
>   nail of the type system.
>
>   I can attempt to simplify and rewrite to the extent that I understand
>   what I'm doing. My own worry is that without direction I might make
>   things worse, or at least not better enough, and I don't want to waste
>   your time with updated pull requests that aren't noticeably better from
>   your perspective.
>
>   Thanks,
>
>   Christophe
>
>   —
>   You are receiving this because you were mentioned.
>   Reply to this email directly, [2]view it on GitHub, or [3]mute the
>   thread.
>
>References
>
>   1. https://github.com/jgm
>   2. https://github.com/jgm/pandoc/pull/2342#issuecomment-230347912
>   3. https://github.com/notifications/unsubscribe/AAAL5Ab8uWuHCzlaxW6MnzI4qsjGKn2Uks5qSWX_gaJpZM4FkKYr

tarleb commented:
I'll try to add a little constructive feedback as well.  Note that this is influenced by my personal coding style.  Kudos for being brave enough and to try and learn. :+1:

One first suggestion would be to look at [this article by Chris Beams](http://chris.beams.io/posts/git-commit/) on how to write a commit message (and why it matters).  It's a very nice guide and describes the de-facto standard for commit messages.

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,6 +1742,41 @@ source = do
+                  xs -> map deunder (drop ((last xs) + 1) s)
+  case elemIndex '|' raw of
+    Nothing -> do
+      let rawstr = ((reverse . (drop 2) . reverse) raw)
```
Most parens here are superfluous.  The outmost pair doesn't add anything, so this becomes

``` haskell
rawstr = (reverse . (drop 2) . reverse) raw
```

Function application has higher precedence than any infix operator, including `(.)`, so the parens around `(drop 2)` can be dropped as well.

``` haskell
rawstr = (reverse . drop 2 . reverse) raw
```

Now finally, one could use the `($)` operator to eliminate the remaining parens:

``` haskell
rawstr = reverse . drop 2 . reverse $ raw
```

That's a matter of taste, though.

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,6 +1742,41 @@ source = do
+      src <- parseFromString (many anyChar) srcstr
+      lab <- parseFromString (trimInlinesF . mconcat <$> many inline) labstr
+      return (lab, src)
+    Just idx -> do
```
The `Nothing` and `Just` cases share a lot of code.  That's usually a sign that things can be abstracted.  One way to do this would be to write a local functions which takes `srcstr` and `labstr` as input and returns `(lab, src)`.

The computation for `rawstr` is the same in both cases, so this should be defined in the previous `let` block.

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,6 +1742,41 @@ source = do
+wikilink :: MarkdownParser (F Inlines)
+wikilink = try $ do
+  guardEnabled Ext_ikiwiki_wikilinks
+  st <- getState
```
Parts of this seem to be copy-pasted from `link`; it would be nicer to extract this in a separate function.

``` haskell
ensureUnnestedLinks :: MarkdownParser a -> MarkdownParser a
```

or something similar.

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,6 +1742,41 @@ source = do
+  (_, raw) <- withRaw $ charsInBalancedBrackets 2
+  guard $ not $ null raw
+  let deunder c = if c == '_' then ' ' else c
+      deiki s = case elemIndices '/' s of
```
Have a look at [`Data.List.span`](http://hackage.haskell.org/package/base-4.9.0.0/docs/Data-List.html#v:span).

csrhodes committed:
tidy up the wikilinks markdown reader

- rename wikilinkstuff to the marginally more descriptive wikilinkInfo
- use span and takeWhile rather than element indexes
- hoist common code into one place as much as possible

csrhodes committed:
tidy up the markdown link writer

here "tidy up" might be in the eye of the beholder.  What I suspect I
really want is a function with guards to handle all the nested if/else
cases, but this compactified if/else chain is the best I could do.

Also use takeWhile rather than elemIndices

csrhodes commented:
Thanks to @tarleb for the comments.  I've tried to address them, and also tried to rework the rightwards-drift in the markdown link writer; again, my lack of Haskell experience is probably showing.  The test suite, and my own application, continue to work with these changes.

tarleb commented:
This looks much better already.  Some issues remain, which I'll comment on inline.

There seem to be merge conflicts, please rebase your branch to the upstream master.

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1475,6 +1475,7 @@ table = try $ do
 
 inline :: MarkdownParser (F Inlines)
 inline = choice [ whitespace
+                , wikilink
```
I like to keep related code grouped together, putting this next to `link` would be more logical.

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,13 +1742,41 @@ source = do
+      rawstr = (reverse . drop 2 . reverse) raw
+      parseLabelAndSrc l s = do
+        lab <- parseFromString (trimInlinesF . mconcat <$> many inline) l
+        src <- parseFromString (many anyChar) s
```
Isn't this a noop (`src == s`)?

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,13 +1742,41 @@ source = do
+  string "[["
+  (_, raw) <- withRaw $ charsInBalancedBrackets 2
+  guard $ not $ null raw
+  let deunder c = if c == '_' then ' ' else c
```
A lot of transformations are performed on the parsed string.  The alternative, which would probably be easier to read, would be to shift this logic into the parsing stage.  It requires a slightly different way to think, but Parsec is a joy to work with.  If you can, give it a try.

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,13 +1742,41 @@ source = do
+  (_, raw) <- withRaw $ charsInBalancedBrackets 2
+  guard $ not $ null raw
+  let deunder c = if c == '_' then ' ' else c
+      deiki s = map deunder $ (reverse . takeWhile (/= '/') . reverse) $ s
```
If you are like me and like [pointfree style](https://wiki.haskell.org/Pointfree)point-less) this could be written as

    deiki = map deunder . reverse . takeWhile (/= '/') . reverse

tarleb commented on src/Text/Pandoc/Readers/Markdown.hs:
```diff
@@ -1741,13 +1742,41 @@ source = do
+  guard $ not $ null raw
+  let deunder c = if c == '_' then ' ' else c
+      deiki s = map deunder $ (reverse . takeWhile (/= '/') . reverse) $ s
+      rawstr = (reverse . drop 2 . reverse) raw
```
Stripping the trailing `]]` is a job probably best done using by Parsec.  See note above.

tarleb commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -956,23 +957,26 @@ inlineToMarkdown opts lnk@(Link attr txt (src, tit))
+  return $
+    let wikilinkString =
+          let deunder c = if c == '_' then ' ' else c
+              deiki s = map deunder $ (reverse . takeWhile (/= '/') . reverse) $ s
```
The `deiki` function is also defined in the reader.  If it's really needes, then maybe this can be put in a central place and imported?  Disclaimer: I haven't checked if it's really needed.

tarleb commented on src/Text/Pandoc/Writers/Markdown.hs:
```diff
@@ -956,23 +957,26 @@ inlineToMarkdown opts lnk@(Link attr txt (src, tit))
+                          then if useShortcutRefLinks then "" else "[]"
+                          else "[" <> reftext <> "]"
+          in  first <> second
+    in if useWikilink then wikilinkString
```
Nice use of lazyness.  I didn't have the time yet to check it for correctness/style, but it sure looks tidy.

tarleb commented:
@csrhodes what are your current thoughts regarding this PR?  Do you plan to continue working on this, or would you rather have somebody pick it up in its present state?  I'll be glad to help either way.

csrhodes commented:
@tarleb unfortunately, I think I'm out of time budget on this for a couple of months - at least, I can probably do small things (I'll try to do the rebase in a moment), but realistically I'm not going to be able to work on e.g. parsing wikilinks with Parsec before around Christmas. I'd be happy to collaborate, if you can spare the time to pick it up; what I really want is to make it easy for committers to commit this to master so that I don't have to maintain a tree - and I understand that the onus is on me to make it easy!

csrhodes committed:
move wikilink in inline to just above link

(just above because [foo] is treated as a reference link, but we want
[[foo]] to be a wikilink if that extension is turned on.)
